jb/itch5/compute_book_depth.cpp

```diff
@@ -0,0 +1,195 @@
+  auto position = orders_.emplace(
+      msg.order_reference_number, order_data{
+        msg.stock, msg.buy_sell_indicator, msg.price, msg.shares});
+  if (position.second == false) {
+    // ... ooops, this should not happen, there is a problem with the
+    // feed, log the problem and skip the message ...
+    auto const& data = position.first->second;
+    JB_LOG(warning) << "duplicate order id=" << msg.order_reference_number
+                    << " existing.symbol=" << data.stock
+                    << ", existing.buy_sell_indicator="
+                    << data.buy_sell_indicator
+                    << ", existing.px=" << data.px
+                    << ", existing.qty=" << data.qty
+                    << ", msg=" << msg;
+    return std::make_tuple(false, order_data(), books_.end());
+  }
+  // ... okay, now that the order is inserted, let's make sure there
+  // is a book for the symbol, we avoid creating a full order book in
+  // the normal case ...
+  auto i = books_.find(msg.stock);
+  if (i == books_.end()) {
+    auto p = books_.emplace(msg.stock, order_book_depth());
+    i = p.first;
+  }
+  // ... add the order to the book and determine if there is an event
+  // ...
+  bool is_event = i->second.handle_add_order(
+      msg.buy_sell_indicator, msg.price, msg.shares);
+  return std::make_tuple(is_event, position.first->second, i);
+}
+
+void jb::itch5::compute_book_depth::handle_reduce(
+    time_point recv_ts, long msgcnt, std::size_t msgoffset,
+    message_header const& header, std::uint64_t order_reference_number,
+    std::uint32_t shares, bool all_shares) {
+  auto r = handle_reduce_no_update(
+      recv_ts, msgcnt, msgoffset, header, order_reference_number,
+      shares, all_shares);
+  if (std::get<0>(r)) {
+    auto const& copy = std::get<1>(r);
+    auto i = std::get<2>(r);
+    // ... if there is an event send that to the callback ...
+    callback_(
+        recv_ts, header, copy.stock,
+        i->second.get_book_depth());
+  }
+}
+
+jb::itch5::compute_book_depth::update_result
+jb::itch5::compute_book_depth::handle_reduce_no_update(
+    time_point recv_ts, long msgcnt, std::size_t msgoffset,
+    message_header const& header, std::uint64_t order_reference_number,
+    std::uint32_t shares, bool all_shares) {
+  // First we need to find the order ...
+  auto position = orders_.find(order_reference_number);
+  if (position == orders_.end()) {
+    // ... ooops, this should not happen, there is a problem with the
+    // feed, log the problem and skip the message ...
+    JB_LOG(warning) << "missing order id=" << order_reference_number
+                    << ", location=" << msgcnt << ":" << msgoffset
+                    << ", header=" << header;
+    return std::make_tuple(false, order_data(), books_.end());
+  }
+
+  // ... okay, now that the order is located, find the book for that
+  // symbol ...
+  auto& data = position->second;
+  auto i = books_.find(data.stock);
```

Comment:
Ugh... 'i' is a horrible name for this variable, 'book' is better (and yes, I realize this is wrong in the original code).  Please consider fixing that.